File: Sample/DaydreamLib/CoreBluetooth+Extensions.m
//
//  CoreBluetooth+Extensions.m
//  DaydreamSample
//
//  Created by David Cairns on 12/29/17.
//  Copyright © 2017 Daydream. All rights reserved.
//

#import "CoreBluetooth+Extensions.h"

@implementation CBService(DCExtensions)

- (CBServiceKind)kind {
    NSString *const uuid = self.UUID.UUIDString;
    if([uuid isEqualToString:@"FE55"]) {
        return CBServiceKindState;
    }
    else if([uuid isEqualToString:@"180F"]) {
        return CBServiceKindBattery;
    }
    else if([uuid isEqualToString:@"180A"]) {
        return CBServiceKindDeviceInfo;
    }
    return CBServiceKindUnknown;
}

@end

@implementation CBCharacteristic(DCExtensions)
- (CBCharacteristicKind)kind {
    NSString *const uuid = self.UUID.UUIDString;
    if([uuid isEqualToString:@"00000001-1000-1000-8000-00805F9B34FB"]) {
        return CBCharacteristicKindState;
    }
    else if([uuid isEqualToString:@"2A29"]) {
        return CBCharacteristicKindManufacturer;
    }
    else if([uuid isEqualToString:@"2A19"]) {
        return CBCharacteristicKindBatteryLevel;
    }
    else if([uuid isEqualToString:@"2A26"]) {
        return CBCharacteristicKindFirmwareVersion;
    }
    else if([uuid isEqualToString:@"2A25"]) {
        return CBCharacteristicKindSerialNumber;
    }
    else if([uuid isEqualToString:@"2A27"]) {
        return CBCharacteristicKindHardwareVersion;
    }
    else if([uuid isEqualToString:@"2A24"]) {
        return CBCharacteristicKindModelNumber;
    }
    else if([uuid isEqualToString:@"2A28"]) {
        return CBCharacteristicKindSoftwareVersion;
    }
    return CBCharacteristicKindUnknown;
}
@end

File: Sources/CoreBluetooth+Extensions.swift
//
//  CoreBluetooth+Extensions.swift
//  DaydreamSample
//
//  Created by David Cairns on 12/28/17.
//  Copyright © 2017 Daydream. All rights reserved.
//

import Foundation
import CoreBluetooth

// MARK: - Convenience
/// A `CBService` extension to easily identify services based on UUID within `DDController` and associated classes.
extension CBService {
    enum Kind {
        case unknown
        case state
        case deviceInfo
        case battery
    }
    
    var kind: CBService.Kind {
        switch uuid.uuidString {
        case "FE55":
            return .state
        case "180F":
            return .battery
        case "180A":
            return .deviceInfo
        default:
            return .unknown
        }
    }
}

/// A `CBCharacteristic` extension to easily identify characteristics based on UUID within `DDController` and associated classes.
extension CBCharacteristic {
    enum Kind: String {
        case unknown
        case state
        case batteryLevel
        case manufacturer
        case firmwareVersion
        case serialNumber
        case hardwareVersion
        case modelNumber
        case softwareVersion
    }
    
    var kind: CBCharacteristic.Kind {
        switch uuid.uuidString {
        case "00000001-1000-1000-8000-00805F9B34FB":
            return .state
        case "2A29":
            return .manufacturer
        case "2A19":
            return .batteryLevel
        case "2A26":
            return .firmwareVersion
        case "2A25":
            return .serialNumber
        case "2A27":
            return .hardwareVersion
        case "2A24":
            return .modelNumber
        case "2A28":
            return .softwareVersion
        default:
            return .unknown
        }
    }
}
